pkg/tfbridge/provider.go

```diff
@@ -954,23 +964,30 @@ func (p *Provider) Update(ctx context.Context, req *pulumirpc.UpdateRequest) (*p
 		diff.SetTimeout(req.Timeout, shim.TimeoutUpdate)
 	}
 
-	newstate, err := p.tf.Apply(res.TFName, state, diff)
-	if newstate == nil {
+	var newstate shim.InstanceState
+	var reasons []string
+	if !req.GetPreview() {
+		newstate, err = p.tf.Apply(res.TFName, state, diff)
+		if newstate == nil {
+			if err != nil {
+				return nil, err
+			}
+
+			return nil, fmt.Errorf("Resource provider reported that the resource did not exist while updating %s.\n\n"+
+				"This is usually a result of the resource having been deleted outside of Pulumi, and can often be "+
+				"fixed by running `pulumi refresh` before updating.", urn)
+		}
+		if newstate.ID() == "" {
+			return nil, fmt.Errorf("expected non-empty ID for new state during Update of %s", urn)
+		}
 		if err != nil {
-			return nil, err
+			reasons = append(reasons, errors.Wrapf(err, "updating %s", urn).Error())
+		}
+	} else {
+		newstate, err = diff.ProposedState(res.TF, nil)
```

Comment:
I haven't spent a lot of time in this code, but given the way this is structured, I would've expected to pass in the resource's prior state here as opposed to `nil` for previewing an update?

Reply:
Indeed, the prior state should be passed in. Still working on tests for this PR, clearly :)

Reply:
Sure thing. Ping me when you're ready for another pass.